rfcs/003 - Multi-Region Support RFC.md

```diff
@@ -0,0 +1,163 @@
+
+I bring this up because it suggests a solution for both problems.
+
+We split the `aws_account` into two tables, `aws_credentials` and `aws_regions`. The credentials table *only* has the credentials for the AWS account, while the `aws_regions` table has three columns, the `region`, `is_default`, and `is_enabled`, with the last two being boolean columns and the `is_default` column having a constraint that only one row has a default value.
+
+When first populating the `aws_regions` table, all regions returned that the user has access to should be enabled by default, and the first one is set as the default. We can then overload the `aws_default_region` function we intend to use for reading the default region to also perform the necessary operations to alter the default region (this requires temporarily disabling the constraint on the table only allowing one `is_default == true` row, making the current default `false`, then setting the new default to `true` and re-enabling the constraint within a transaction).
+
+Within the `AwsAccount` object in the engine itself, it currently attaches a `getAwsClient` method to the context object that the other modules use. This should continue to maintain the current behavior it has when no arguments are passed to it, returning a client tied to the default region, which will allow existing modules to be converted to region-aware over time.
+
+There should also be a new function to `getEnabledAwsRegions` and `getAwsClient` should be overloaded to accept a region argument to override the default.
+
+The cloud read functions would need to be updated to first `getEnabledAwsRegions` and then loop their current logic per region received. Cloud create/update/delete would use their own entity's `region` column to determine the value to pass to the `getAwsClient` function.
+
+Depending on how the AWS API rate limiting works, we will either need to execute the cloud reads sequentially across all enabled regions, or we could potentially fire them off in parallel and negate most of the expected performance impact of going multi-region. Because this approach has region logic fully contained within the tables/entities, we can use either solution without altering the schema exposed to the end user, so we can experiment with parallelization without an issue; otherwise minimizing performance degradation would be handled via `getEnabledAwsRegions` and the user eliminating all regions they do not wish to manage.
+
+### Alternatives Considered
+
+#### `region` columns on all relevant tables
+
+The simplest solution is to just slap a `region` column onto every table that is region-specific and cloud reads cycle through all regions every time.
+
+##### Pros
+
+1. We can implement this piecemeal, module-by-module ignoring the `region` column in the `aws_account` and doing their own thing.
+2. Very clear where resources are located.
```

Comment:
@dfellis in your main solution, how to determine which region a resource is located in?

Reply:
Oh, if something is region-specific it will have a region column. The columns will all get their default value from a function that returns their chosen default, so there is not change necessary for any `insert` so only users that go multiregion will have the added complexity.